Assets/MixedRealityToolkit/Providers/Hands/BaseHand.cs

```diff
@@ -54,30 +56,39 @@ public override void SetupDefaultInteractions(Handedness controllerHandedness)
 
         protected void UpdateVelocity()
         {
-            if (frameOn == 0)
+            if(frameOn < velocityUpdateInterval)
             {
-                deltaTimeStart = Time.unscaledTime;
-                lastPosition = GetJointPosition(TrackedHandJoint.Palm);
-                lastPalmNormal = GetPalmNormal();
-                //lastPalmNormal = new Vector3(1, 0, 0);
+                velocityPositionsCache[frameOn] = GetJointPosition(TrackedHandJoint.Palm);
+                velocityPositionsSum += velocityPositionsCache[frameOn];
+                velocityNormalsCache[frameOn] = GetPalmNormal();
+                velocityNormalsSum += velocityNormalsCache[frameOn];
             }
-            else if (frameOn == velocityUpdateInterval)
+            else
             {
-                //update linear velocity
+                int frameIndex = frameOn % velocityUpdateInterval;
+
                 float deltaTime = Time.unscaledTime - deltaTimeStart;
-                Vector3 newVelocity = (GetJointPosition(TrackedHandJoint.Palm) - lastPosition) / deltaTime;
-                Velocity = (Velocity * 0.8f) + (newVelocity * 0.2f);
 
-                //update angular velocity
-                Vector3 currentPalmNormal = GetPalmNormal();
-                //currentPalmNormal = new Vector3(0, 1, 0);
-                Quaternion rotation = Quaternion.FromToRotation(lastPalmNormal, currentPalmNormal);
+                Vector3 newPosition = GetJointPosition(TrackedHandJoint.Palm);
+                Vector3 newNormal = GetPalmNormal();
+
+                Vector3 newPositionsSum = velocityPositionsSum - velocityPositionsCache[frameIndex] + newPosition;
+                Vector3 newNormalsSum = velocityNormalsSum - velocityNormalsCache[frameIndex] + newNormal;
+
+                Velocity = (newPositionsSum - velocityPositionsSum) / deltaTime / velocityUpdateInterval;
+
+                Quaternion rotation = Quaternion.FromToRotation(velocityNormalsSum / velocityUpdateInterval, newNormalsSum / velocityUpdateInterval);
                 Vector3 rotationRate = rotation.eulerAngles * Mathf.Deg2Rad;
                 AngularVelocity = rotationRate / deltaTime;
+
+                velocityPositionsCache[frameIndex] = newPosition;
+                velocityNormalsCache[frameIndex] = newNormal;
+                velocityPositionsSum = newPositionsSum;
+                velocityNormalsSum = newNormalsSum;
```

Comment:
This velocity calculation code looks to be decently common between a few different files - it might be worth abstracting that out into a helper class (i.e. VelocityCalculator.cs), where each frame you're responsible for calling UpdateVelocity(NewPosition, NewNormal), and then the caller can then just call GetVelocity and GetAngularVelocity.

This would also make it possible to unit test this functionality (i.e. you could give it 10 different points and normal and then show the expected numerical amount)

Reply:
I wasn't sure if we would want the same algorithm for all controllers, because robustness of positions varies from controller to controller. So I chose not to abstract it yet. I can do that if we think this proposed implementation is okay for the foreseeable future.

Reply:
I changed the other 2 to platform values, so this code is only used for simulated hands now. Seems okay to leave it as is.